File: code/uva/540.cpp
#include <bits/stdc++.h>
using namespace std;
const int N = 1005;
map<int, int> mp;
queue<int> que[N];
queue<int> rec;
int main()
{
	int t;
	int cas = 1;
	while (~scanf("%d", &t) && t) {
		printf("Scenario #%d\n", cas++);
		while (!rec.empty())
			rec.pop();
		for (int i = 1; i <= t; i++) {
			while (!que[i].empty())
				que[i].pop();
			int x, y;
			scanf("%d", &x);
			for (int j = 0; j < x; j++) {
				scanf("%d", &y);
				mp[y] = i;
			}
		}
		char str[20];
		while (~scanf("%s", str) && str[0] != 'S') {
			if (str[0] == 'E') {
				int x;
				scanf("%d", &x);
				if (que[mp[x]].empty()) {
					rec.push(mp[x]);
				}
				que[mp[x]].push(x);
			} else {
				int tmp = rec.front();
				printf("%d\n", que[tmp].front());
				que[tmp].pop();
				if (que[tmp].empty()) {
					rec.pop();
				}
			}
		}
		putchar('\n');
	}
	return 0;
}

File: code/uva/540.py
import Queue
N = 1005
que = Queue.Queue()
rec[1005] = Queue.Queue()
dict = {};
cas = 1
while True:
	try:
		t = int(input())
		print("Scenario #", cas)
		cas += 1
		for i in range(t):
			n = int(input())
			for j in range(n):
				x = int(input())
				dict[x] = i
		while True:
			str = input()
			if str == "ENQUEUE":
				x = int(input())
				if (!rec[dict[x]].empty()):
					que.put(dict[x])
					rec[x].put(x)
				else:
					rec[x].put(x)
					que.put(dict)

			elif str == "DEQUEUE":
				now = que.get()
				print(rec[now].get())
			else:
				break
	except EOFError:
		break